src/cargo-kani/src/metadata.rs

```diff
@@ -0,0 +1,193 @@
+fn merge_kani_metadata(files: Vec<KaniMetadata>) -> KaniMetadata {
+    let mut result = KaniMetadata { proof_harnesses: Vec::new() };
+    for md in files {
+        // Note that we're taking ownership of the original vec, and so we can move the data into the new data structure.
+        result.proof_harnesses.extend(md.proof_harnesses);
+    }
+    result
+}
+
+impl KaniSession {
+    /// Reads a collection of kani-metadata.json files and merges the results.
+    pub fn collect_kani_metadata(&self, files: &[PathBuf]) -> Result<KaniMetadata> {
+        if self.args.dry_run {
+            // Mock an answer
+            Ok(KaniMetadata { proof_harnesses: vec![] })
+        } else {
+            // TODO: one possible future improvement here would be to return some kind of Lazy
+            // value, that only computes this metadata it turns out we need it.
+            let results: Result<Vec<_>, _> = files.iter().map(|x| read_kani_metadata(x)).collect();
+            Ok(merge_kani_metadata(results?))
+        }
+    }
+
+    /// Determine which function to use as entry point, based on command-line arguments and kani-metadata.
+    pub fn determine_targets(&self, metadata: &KaniMetadata) -> Result<Vec<HarnessMetadata>> {
+        if let Some(name) = &self.args.function {
+            // --function is untranslated, create a mock harness
+            return Ok(vec![mock_proof_harness(name)]);
+        }
+        if let Some(name) = &self.args.harness {
+            // Linear search, since this is only ever called once
+            if let Some(harness) = find_proof_harness(name, &metadata.proof_harnesses) {
+                return Ok(vec![harness.clone()]);
+            }
+            bail!("A proof harness named {} was not found", name);
+        }
+        if metadata.proof_harnesses.is_empty() {
+            // TODO: This could use a better error message, possibly with links to Kani documentation.
+            // New users may encounter this and could use a pointer to how to write proof harnesses.
+            bail!("No proof harnesses (functions with #[kani::proof]) were found to verify.");
+        } else {
+            Ok(metadata.proof_harnesses.clone())
+        }
+    }
+}
+
+fn mock_proof_harness(name: &str) -> HarnessMetadata {
+    HarnessMetadata {
+        pretty_name: name.into(),
+        mangled_name: name.into(),
+        original_file: "<unknown>".into(),
+        original_line: "<unknown>".into(),
+        unwind_value: None,
+    }
+}
+
+fn find_proof_harness<'a>(
+    name: &str,
+    harnesses: &'a [HarnessMetadata],
+) -> Option<&'a HarnessMetadata> {
+    for h in harnesses.iter() {
+        if h.pretty_name == *name {
+            return Some(h);
+        } else {
+            // pretty_name will be things like `module::submodule::name_of_function`
+            // and we want people to be able to specify `--harness name_of_function`
+            if let Some(prefix) = h.pretty_name.strip_suffix(name) {
```

Comment:
We should probably check for the case where there is ambiguity, i.e., multiple functions that have the same name, e.g. `module1::submodule1::name_of_function` and `module2::submodule2::name_of_function`, and emit an error.